### PR TITLE
Support for 32-bit floats

### PR DESCRIPTION
**Pull request by garth-wells**
This adds support and testing for 32-bit float meshes and Functions.

There is more to do still on mesh IO and a small number of tests that do not pass with `float32` (disabled for now with `float32`).

Testing for `std::complex<float>` will be added later.

Note change to Python BoundingBoxTree creation to handle different float types.

### CODE REVIEW

File: python/dolfinx/io/utils.py [hunk 1]
```diff
@@ -38,7 +38,8 @@ def _extract_cpp_functions(functions: typing.Union[typing.List[Function], Functi
     # FidesWriter and VTXWriter require ADIOS2
     __all__ = __all__ + ["FidesWriter", "VTXWriter"]
 
-    class VTXWriter(_cpp.io.VTXWriter):
+    class VTXWriter:
+        # class VTXWriter(_cpp.io.VTXWriter):
```
Comment (chrisrichardson) [inline python/dolfinx/io/utils.py:42]: remove commented line

File: python/dolfinx/wrappers/assemble.cpp [hunk 2]
```diff
@@ -380,86 +467,11 @@ void petsc_module(py::module& m)
...
-        // Build operator
-        Mat A = dolfinx::la::petsc::create_matrix(comm, sp);
-        MatSetOption(A, MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE);
-        dolfinx::fem::interpolation_matrix<PetscScalar>(
-            V0, V1, dolfinx::la::petsc::Matrix::set_block_fn(A, INSERT_VALUES));
-        return A;
-      },
-      py::return_value_policy::take_ownership, py::arg("V0"), py::arg("V1"));
+  declare_petsc_discrete_operators<PetscScalar, PetscReal>(m);
+  // declare_petsc_discrete_operators<float, float>(m);
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/assemble.cpp:471]: Remove commented lines

File: python/dolfinx/wrappers/io.cpp [hunk 3]
```diff
@@ -26,11 +26,58 @@
 #include <string>
 #include <vector>
 
+#include <petscsystypes.h>
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/io.cpp:29]: I find this objectionable - I/O has nothing to do with PETSc. Can't this be fixed with a little more templating?

File: python/test/unit/fem/test_custom_assembler.py [hunk 4]
```diff
@@ -296,21 +295,21 @@ def test_custom_mesh_loop_rank1():
...
-    # for i in range(2):
-    #     b = btmp.x.array
-    #     b[:] = 0.0
-    #     start = time.time()
-    #     assemble_vector_parallel(b, x_dofs, x, dofmap_t.array, dofmap_t.offsets, num_owned_cells)
-    #     end = time.time()
-    #     print("Time (numba parallel, pass {}): {}".format(i, end - start))
-    # btmp.vector.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
-    # assert (btmp.vector - b0.vector).norm() == pytest.approx(0.0)
+    # # NOTE: Parallel (threaded) Numba can cause problems with MPI
```
Comment (chrisrichardson) [inline python/test/unit/fem/test_custom_assembler.py:298]: Remove double commenting

File: python/test/unit/fem/test_custom_assembler.py [hunk 5]
```diff
@@ -331,12 +330,32 @@ def test_custom_mesh_loop_rank1():
     assert (b1 - b0.vector).norm() == pytest.approx(0.0)
 
     # Assemble using generated tabulate_tensor kernel and Numba assembler
-    ffcxtype = "double _Complex" if np.issubdtype(PETSc.ScalarType, np.complexfloating) else "double"
+    # ffcxtype = "double _Complex" if np.issubdtype(PETSc.ScalarType, np.complexfloating) else "double"
```
Comment (chrisrichardson) [inline python/test/unit/fem/test_custom_assembler.py:333]: remove commented line

File: python/test/unit/fem/test_fem_pipeline.py [hunk 6]
```diff
@@ -270,6 +270,7 @@ def test_curl_curl_eigenvalue(family, order):
     B.destroy()
 
 
+# @pytest.mark.skipif(default_real_type != np.float64, reason="float32 not supported yet")
```
Comment (chrisrichardson) [inline python/test/unit/fem/test_fem_pipeline.py:273]: commented line

File: python/test/unit/fem/test_element_integrals.py [hunk 7]
```diff
@@ -131,8 +131,8 @@ def two_unit_cells(cell_type, agree=False, random_order=True, return_order=False
     return mesh
 
 
-@pytest.mark.skip_in_parallel
-@parametrize_cell_types
+@ pytest.mark.skip_in_parallel
```
Comment (chrisrichardson) [inline python/test/unit/fem/test_element_integrals.py:134]: Why `@ ` - this doesn't seem standard elsewhere

File: python/test/unit/fem/test_function.py [hunk 8]
```diff
@@ -175,11 +177,13 @@ def f(x):
     assert round(w.vector.norm(PETSc.NormType.N1) - 3 * num_vertices, 7) == 0
 
 
+# @pytest.mark.skipif(default_real_type != np.float64, reason="float32 not supported yet")
```
Comment (chrisrichardson) [inline python/test/unit/fem/test_function.py:180]: remove commented line

**Review by jorgensd — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/utils.h [hunk 9]
```diff
@@ -1045,16 +1045,11 @@ Expression<T, U> create_expression(
                              "function space was provided.");
   }
 
-  std::span<const U> X(e.points, e.num_points * e.topological_dimension);
+  std::vector<U> X(e.points, e.points + e.num_points * e.topological_dimension);
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/utils.h:1048]: Yes, but one step at a time. This is already a big change.

File: python/dolfinx/geometry.py [hunk 10]
```diff
@@ -16,62 +16,146 @@
...
-            raise RuntimeError(f"Mesh entities of dimension {dim} have not been created.")
-        if entities is None:
-            entities = range(0, map.size_local + map.num_ghosts)
+        return self._cpp_object.get_bbox(i)
+
+    def create_global_tree(self, comm) -> BoundingBoxTree:
+        return BoundingBoxTree(self._cpp_object.create_global_tree(comm))
+
+
+def bb_tree(mesh: Mesh, dim: int, entities=None, padding: float = 0.0) -> BoundingBoxTree:
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/geometry.py]: Type hint for padding is fine as it is. Value will be cast if needed.

File: cpp/demo/hyperelasticity/main.cpp [hunk 11]
```diff
@@ -182,7 +182,7 @@ int main(int argc, char* argv[])
         *V,
         [](auto x)
         {
-          constexpr double eps = 1.0e-8;
+          constexpr U eps = 1.0e-6;
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/demo/hyperelasticity/main.cpp:185]: Single precision is precise to about 7 decimal digits, so 10^-8 can be a bit tight.

File: python/dolfinx/wrappers/io.cpp [hunk 12]
```diff
@@ -123,22 +184,37 @@ void io(py::module& m)
           [](dolfinx::io::XDMFFile& self, std::string name, std::string xpath)
           {
             auto [x, shape] = self.read_geometry_data(name, xpath);
-            return as_pyarray(std::move(x), shape);
+            std::vector<double>& _x = std::get<std::vector<double>>(x);
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/wrappers/io.cpp:187]: Mesh input is the next step.

File: python/dolfinx/wrappers/io.cpp (hunk 3 above)
Comment (garth-wells) [reply python/dolfinx/wrappers/io.cpp]: Fixed.

File: python/dolfinx/wrappers/geometry.cpp [hunk 13]
```diff
@@ -20,30 +20,94 @@
...
 {
+  // dolfinx::geometry::BoundingBoxTree
+  std::string pyclass_name = "BoundingBoxTree_" + type;
+  py::class_<dolfinx::geometry::BoundingBoxTree<T>,
+             std::shared_ptr<dolfinx::geometry::BoundingBoxTree<T>>>(
+      m, pyclass_name.c_str())
+      .def(py::init(
+               [](const dolfinx::mesh::Mesh<T>& mesh, int dim,
+                  const py::array_t<std::int32_t, py::array::c_style>& entities,
+                  double padding)
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/wrappers/geometry.cpp:36]: No need - cast to double is fine here.

File: python/test/unit/fem/test_assemble_domains.py [hunk 14]
```diff
@@ -93,11 +93,11 @@ def test_assembly_dx_domains(mode, meshtags_factory):
     L = form(w * (dx(1) + dx(2) + dx(3)))
     s = assemble_scalar(L)
     s = mesh.comm.allreduce(s, op=MPI.SUM)
-    assert s == pytest.approx(0.5, 1.0e-12)
+    assert s == pytest.approx(0.5, rel=1.0e-6)
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/test/unit/fem/test_assemble_domains.py:96]: Ideally less, but hard to tune for the general case. Let's look at this later.

File: python/test/unit/fem/test_assemble_submesh.py [hunk 15]
```diff
@@ -105,8 +105,8 @@ def test_submesh_facet_assembly(n, k, space, ghost_mode):
     square_mesh = create_unit_square(MPI.COMM_WORLD, n, n, ghost_mode=ghost_mode)
     A_square_mesh, b_square_mesh, s_square_mesh = assemble(square_mesh, space, k)
 
-    assert np.isclose(A_submesh.norm(), A_square_mesh.norm())
-    assert np.isclose(b_submesh.norm(), b_square_mesh.norm())
+    assert A_submesh.norm() == pytest.approx(A_square_mesh.norm(), rel=1.0e-5, abs=1.0e-5)
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/test/unit/fem/test_assemble_submesh.py:108]: We're testing with `pytest`, so neater to not introduce another library for tested where it's not needed.

**Review by jorgensd — APPROVED** [review]

**Pull request merged**